implementations/rust/vault/src/c/mod.rs

```diff
@@ -1,20 +1,771 @@
+use std::ffi::CStr;
 use zeroize::Zeroize;
 
+/// C vault secret
+#[derive(Debug)]
+pub struct CVaultSecret {
+    handle: ockam_vault_secret_t,
+}
+
+// FIXME
+unsafe impl Send for CVaultSecret {}
+
+// FIXME
+unsafe impl Sync for CVaultSecret {}
+
+impl Drop for CVaultSecret {
+    fn drop(&mut self) {
+        // FIXME
+    }
+}
+
+impl Secret for CVaultSecret {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_mut_any(&mut self) -> &mut dyn Any {
+        self
+    }
+}
+
+/// Builder
+#[derive(Debug)]
+pub struct Atecc608aVaultBuilder {}
+
+impl Default for Atecc608aVaultBuilder {
+    fn default() -> Self {
+        Atecc608aVaultBuilder {}
+    }
+}
+
+impl Atecc608aVaultBuilder {
+    /// Build
+    pub fn build(self) -> Result<CVault, VaultFailError> {
+        let mut c_vault = ockam_vault_t {
+            dispatch: std::ptr::null_mut(),
+            context: std::ptr::null_mut(),
+        };
+
+        let memory = RustAlloc::new();
+
+        let t = ATCAIfaceCfg__bindgen_ty_1 {
+            atcai2c: ATCAIfaceCfg__bindgen_ty_1__bindgen_ty_1 {
+                slave_address: 0xC0,
+                bus: 1,
+                baud: 100000,
+            },
+        };
+
+        let mut atca_cfg = ATCAIfaceCfg {
+            iface_type: ATCAIfaceType::ATCA_I2C_IFACE,
+            devtype: ATCADeviceType::ATECC608A,
+            __bindgen_anon_1: t,
+            wake_delay: 1500,
+            rx_retries: 20,
+            cfg_data: std::ptr::null_mut(),
+        };
+
+        let mut io_protection = ockam_vault_atecc608a_io_protection_t {
+            /* IO Protection Key is used to encrypt data sent via */
+            key: [
+                0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06,
+                0x07, /* I2C to the ATECC608A. During init the key is       */
+                0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16,
+                0x17, /* written into the device. In a production system    */
+                0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26,
+                0x27, /* the key should be locked into the device and never */
+                0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36,
+                0x37, /* transmitted via I2C.                               */
+            ],
+            key_size: 32,
+            slot: 6,
+        };
+
+        let mut attributes = ockam_vault_atecc608a_attributes_t {
+            memory: memory.as_mut_ptr(),
+            mutex: std::ptr::null_mut(),
+            atca_iface_cfg: &mut atca_cfg,
+            io_protection: &mut io_protection,
+        };
+
+        let error = unsafe { ockam_vault_atecc608a_init(&mut c_vault, &mut attributes) };
+        CVault::handle_error(error)?;
+
+        Ok(CVault::new(c_vault))
+    }
+}
+
 // TODO: Should be thread-safe?
 /// Represents a single instance of an Ockam vault context
 #[derive(Debug)]
 pub struct CVault {
-    context: Cell<ockam_vault_t>,
+    context: ockam_vault_t,
+}
+
+impl Drop for CVault {
```

Comment:
I don't think the drop is necessary since it's not deleting anything and just a reference pointer.

Reply:
Yes, we have Vault::deinit instead. But I'm not sure we use it...

Reply:
Revisit later